src/native/http_request_utils.c

```diff
@@ -310,126 +322,89 @@ struct aws_http_message *aws_http_request_new_from_java_http_request(
     return request;
 
 on_error:
+    if (exception_message) {
+        aws_jni_throw_runtime_exception(env, exception_message);
+    }
 
     /* Don't need to destroy input stream since it's the last thing created */
     aws_http_message_destroy(request);
 
     return NULL;
 }
 
-jobjectArray aws_java_headers_array_from_native(
-    JNIEnv *env,
-    const struct aws_http_header *header_array,
-    size_t num_headers) {
-
-    jobjectArray jArray = (*env)->NewObjectArray(env, (jsize)num_headers, http_header_properties.header_class, NULL);
+static inline int s_marshall_http_request(const struct aws_http_message *message, struct aws_byte_buf *request_buf) {
+    struct aws_byte_cursor method;
+    AWS_ZERO_STRUCT(method);
 
-    for (size_t i = 0; i < num_headers; i++) {
-        jobject jHeader =
-            (*env)->NewObject(env, http_header_properties.header_class, http_header_properties.constructor);
+    AWS_FATAL_ASSERT(!aws_http_message_get_request_method(message, &method));
 
-        jbyteArray actual_name = aws_jni_byte_array_from_cursor(env, &(header_array[i].name));
-        jbyteArray actual_value = aws_jni_byte_array_from_cursor(env, &(header_array[i].value));
+    struct aws_byte_cursor path;
+    AWS_ZERO_STRUCT(path);
 
-        // Overwrite with actual values
-        (*env)->SetObjectField(env, jHeader, http_header_properties.name, actual_name);
-        (*env)->SetObjectField(env, jHeader, http_header_properties.value, actual_value);
-        (*env)->SetObjectArrayElement(env, jArray, (jsize)i, jHeader);
+    AWS_FATAL_ASSERT(!aws_http_message_get_request_path(message, &path));
 
-        (*env)->DeleteLocalRef(env, actual_name);
-        (*env)->DeleteLocalRef(env, actual_value);
-        (*env)->DeleteLocalRef(env, jHeader);
+    if (aws_byte_buf_reserve_relative(request_buf, sizeof(int) + sizeof(int) + method.len + path.len)) {
```

Comment:
Is it worth adding some arbitrary padding here to avoid the realloc that will inevitably happen because everything has a header?

Reply:
hopefully we're doing logarithmic growth on the buffers. are we?

Reply:
Sure, but I was thinking something along the lines "90% of the time header size is below N (1K?)" so just add 1K padding and then 90% of the time there's no realloc at all.  I guess the question is "is there distribution-based reasoning/data here that we can apply to reduce the allocations from 2 to 1 the majority of the time?"

Reply:
oh, sorry, missed this, We already do EXACTLY this on line 376, that line is just a precautionary  call, to make sure we have enough space.